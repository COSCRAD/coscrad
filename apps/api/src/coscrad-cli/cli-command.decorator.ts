/**
 * Given that we have an extensive custom set of decorators for commands in the
 * CQRS sense of the word "command", we re-export and alias `nest-commander` decorators
 * for clarity in naming. These decorators should be imported from this local file
 * into any CLI Command files.
 */
export {
    Command as CliCommand,
    CommandRunner as CliCommandRunner,
    Option as CliCommandOption,
} from 'nest-commander';
