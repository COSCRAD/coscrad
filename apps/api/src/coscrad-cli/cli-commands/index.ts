export * from './performance-test.cli-command';
