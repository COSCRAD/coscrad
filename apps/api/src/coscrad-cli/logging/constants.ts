export const COSCRAD_CLI_LOGGER_TOKEN = `COSCRAD_CLI_LOGGER_TOKEN`;
