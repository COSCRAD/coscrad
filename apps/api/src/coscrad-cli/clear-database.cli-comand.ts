import { Inject } from '@nestjs/common';
import { ArangoDatabaseProvider } from '../persistence/database/database.provider';
import { DataImporter } from '../persistence/repositories/data-importer';
import { CliCommand, CliCommandRunner } from './cli-command.decorator';
import { COSCRAD_LOGGER_TOKEN, ICoscradLogger } from './logging';

@CliCommand({
    name: 'clear-database',
    description: '(Cypress mode only) clears database for automated tests',
})
export class ClearDatabaseCliCommand extends CliCommandRunner {
    dataImporter: DataImporter;

    constructor(
        databaseProvider: ArangoDatabaseProvider,
        @Inject(COSCRAD_LOGGER_TOKEN) private readonly logger: ICoscradLogger
    ) {
        super();

        this.dataImporter = new DataImporter(databaseProvider, this.logger);
    }

    // note that this will not work unless in a testing environment with `$DATA_MODE=_CYPRESS`
    async run(_passedParams: string[]) {
        await this.dataImporter.deleteAllData();

        this.logger.log(`Successfully emptied all collections`);
    }
}
