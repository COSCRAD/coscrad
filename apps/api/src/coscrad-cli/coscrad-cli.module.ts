import { CommandModule } from '@coscrad/commands';
import { Module } from '@nestjs/common';
import { AppModule } from '../app/app.module';
import { EdgeConnectionModule } from '../app/domain-modules/edge-connection.module';
import { MediaItemModule } from '../app/domain-modules/media-item.module';
import { VocabularyListModule } from '../app/domain-modules/vocabulary-list.module';
import { EventModule } from '../domain/common';
import { AudioVisualModule } from '../domain/models/audio-visual/application/audio-visual.module';
import { IdGenerationModule } from '../lib/id-generation/id-generation.module';
import { MigrationModule } from '../persistence/migrations';
import { PersistenceModule } from '../persistence/persistence.module';
import { ClearDatabaseCliCommand } from './clear-database.cli-comand';
import { DomainDumpCliCommand } from './data-dump.cli-command';
import { DomainRestoreCliCommand } from './data-restore.cli-command';
import { ExecuteCommandStreamCliCommand } from './execute-command-stream.cli-command';
import { ExportAudioItemLineagesCliCommand } from './export-audio-item-lineages.cli-command';
import { ExportMediaAnnotationsCliCommand } from './export-media-annotations.cli-command';
import { ExportSchemasCliCommand } from './export-schemas.cli-command';
import { IngestMediaItemsCliCommand } from './ingest-media-items.cli-command';
import { ListMigrationsCliCommand } from './list-migrations.cli-command';
import { ConsoleCoscradCliLogger, COSCRAD_LOGGER_TOKEN } from './logging';
import { RehydrateViewsCliCommand } from './rehydrate-views.cli-command';
import { RevertLatestMigrationCliCommand } from './revert-latest-migration';
import { RunMigrationsCliCommand } from './run-migrations.cli-command';
import { SeedDatabaseCliCommand } from './seed-database.cli-command';
import { SeedTestDataWithCommand } from './seed-test-data-with-command.cli-command';
import { SeedTestUuids } from './seed-test-uuids.cli-command';
import { ValidateInvariantsCliCommand } from './validate-invariants.cli-command';

@Module({
    providers: [
        DomainDumpCliCommand,
        DomainRestoreCliCommand,
        ListMigrationsCliCommand,
        RunMigrationsCliCommand,
        RevertLatestMigrationCliCommand,
        ValidateInvariantsCliCommand,
        ClearDatabaseCliCommand,
        ExecuteCommandStreamCliCommand,
        SeedDatabaseCliCommand,
        SeedTestDataWithCommand,
        SeedTestUuids,
        IngestMediaItemsCliCommand,
        ExportMediaAnnotationsCliCommand,
        ExportAudioItemLineagesCliCommand,
        ExportSchemasCliCommand,
        RehydrateViewsCliCommand,
        {
            provide: COSCRAD_LOGGER_TOKEN,
            useClass: ConsoleCoscradCliLogger,
        },
    ],
    imports: [
        AppModule,
        PersistenceModule,
        IdGenerationModule,
        MigrationModule,
        CommandModule,
        VocabularyListModule,
        AudioVisualModule,
        MediaItemModule,
        EdgeConnectionModule,
        EventModule,
    ],
})
export class CoscradCliModule {}
