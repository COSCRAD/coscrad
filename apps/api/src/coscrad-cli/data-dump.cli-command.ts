import { Inject } from '@nestjs/common';
import { writeFileSync } from 'fs';
import { IRepositoryProvider } from '../domain/repositories/interfaces/repository-provider.interface';
import { REPOSITORY_PROVIDER_TOKEN } from '../persistence/constants/persistenceConstants';
import { DomainDataExporter } from '../persistence/repositories/domain-data-exporter';
import convertInMemorySnapshotToDatabaseFormat from '../test-data/utilities/convertInMemorySnapshotToDatabaseFormat';
import { CliCommand, CliCommandOption, CliCommandRunner } from './cli-command.decorator';

@CliCommand({
    name: 'data-dump',
    description: 'dumps the database state to a snapshot file',
})
export class DomainDumpCliCommand extends CliCommandRunner {
    private readonly dataExporter: DomainDataExporter;

    constructor(@Inject(REPOSITORY_PROVIDER_TOKEN) repositoryProvider: IRepositoryProvider) {
        super();

        this.dataExporter = new DomainDataExporter(repositoryProvider);
    }

    async run(_passedParams: string[], options?: { filepath: string }): Promise<void> {
        const NUMBER_OF_SPACES_TO_INDENT = 4;

        const domainSnapshot = await this.dataExporter.fetchSnapshot();

        const snapshotInDatabaseFormat = convertInMemorySnapshotToDatabaseFormat(
            domainSnapshot.fetchFullSnapshotInLegacyFormat()
        );

        writeFileSync(
            options.filepath,
            JSON.stringify(snapshotInDatabaseFormat, null, NUMBER_OF_SPACES_TO_INDENT)
        );
    }

    @CliCommandOption({
        flags: '-f, --filepath [filepath]',
        description: 'the path to write the output to',
        required: true,
    })
    parseFilepath(value: string): string {
        return value;
    }
}
