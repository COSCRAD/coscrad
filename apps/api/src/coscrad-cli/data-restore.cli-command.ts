import { Inject } from '@nestjs/common';
import { readFileSync } from 'fs';
import { InternalError } from '../lib/errors/InternalError';
import { ArangoQueryRunner } from '../persistence/database/arango-query-runner';
import { ArangoDatabaseProvider } from '../persistence/database/database.provider';
import { ArangoDataExporter } from '../persistence/repositories/arango-data-exporter';
import { InMemoryDatabaseSnapshot } from '../test-data/utilities';
import { CliCommand, CliCommandOption, CliCommandRunner } from './cli-command.decorator';
import { COSCRAD_LOGGER_TOKEN, ICoscradLogger } from './logging';

@CliCommand({
    name: 'data-restore',
    description: 'restores the database state from a snapshot file',
})
export class DomainRestoreCliCommand extends CliCommandRunner {
    private readonly dataExporter: ArangoDataExporter;

    constructor(
        databaseProvider: ArangoDatabaseProvider,
        @Inject(COSCRAD_LOGGER_TOKEN) private readonly logger: ICoscradLogger
    ) {
        super();

        this.dataExporter = new ArangoDataExporter(new ArangoQueryRunner(databaseProvider));
    }

    async run(_passedParams: string[], { filepath }: { filepath: string }): Promise<void> {
        const parsedSnapshot = JSON.parse(
            readFileSync(filepath, { encoding: 'utf-8' })
        ) as InMemoryDatabaseSnapshot;

        await this.dataExporter.restoreFromSnapshot(parsedSnapshot).catch((e) => {
            const error = new InternalError(`Data exporter failed to restore data.`, [e]);

            this.logger.log(error.toString());

            throw error;
        });
    }

    @CliCommandOption({
        flags: '-f, --filepath [filepath]',
        description: 'the path to write the output to',
        required: true,
    })
    parseFilepath(value: string): string {
        return value;
    }
}
