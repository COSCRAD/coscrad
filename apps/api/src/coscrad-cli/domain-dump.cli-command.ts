import { Inject } from '@nestjs/common';
import { writeFileSync } from 'fs';
import { Command, CommandRunner, Option } from 'nest-commander';
import { IRepositoryProvider } from '../domain/repositories/interfaces/repository-provider.interface';
import { ResourceType } from '../domain/types/ResourceType';
import { REPOSITORY_PROVIDER_TOKEN } from '../persistence/constants/persistenceConstants';

@Command({
    name: 'domain-dump',
    description: 'test the COSCRAD CLI!',
})
export class DomainDumpCliCommand extends CommandRunner {
    constructor(
        @Inject(REPOSITORY_PROVIDER_TOKEN) private readonly repositoryProvider: IRepositoryProvider
    ) {
        super();
    }

    async run(_passedParams: string[], options?: Record<string, any>): Promise<void> {
        const result = await this.repositoryProvider.forResource(ResourceType.term).fetchMany();

        writeFileSync(options.filepath, JSON.stringify(result, null, 4));

        Promise.resolve();
    }

    @Option({
        flags: '-f, --filepath [filepath]',
        description: 'the path to write the output to',
        required: true,
    })
    parseFilepath(value: string): string {
        return value;
    }
}
