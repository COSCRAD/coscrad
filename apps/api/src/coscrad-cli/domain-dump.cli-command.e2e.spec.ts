import { CommandTestFactory } from 'nest-commander-testing';
import { CoscradCliModule } from './coscrad-cli.module';

import { TestingModule } from '@nestjs/testing';
import { existsSync, readFileSync, unlinkSync } from 'fs';
import { AppModule } from '../app/app.module';
import createTestModule from '../app/controllers/__tests__/createTestModule';
import { DeluxeInMemoryStore } from '../domain/types/DeluxeInMemoryStore';
import { REPOSITORY_PROVIDER_TOKEN } from '../persistence/constants/persistenceConstants';
import { ArangoConnectionProvider } from '../persistence/database/arango-connection.provider';
import { ArangoDatabaseProvider } from '../persistence/database/database.provider';
import generateDatabaseNameForTestSuite from '../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import TestRepositoryProvider from '../persistence/repositories/__tests__/TestRepositoryProvider';
import buildTestDataInFlatFormat from '../test-data/buildTestDataInFlatFormat';

describe('Task Command', () => {
    const outputFilePath = `__cli-command-test-files__/domain-dump.spec.data.json`;

    let commandInstance: TestingModule;

    let testRepositoryProvider: TestRepositoryProvider;

    beforeAll(async () => {
        const testAppModule = await createTestModule({
            ARANGO_DB_NAME: generateDatabaseNameForTestSuite(),
        });

        const arangoConnectionProvider =
            testAppModule.get<ArangoConnectionProvider>(ArangoConnectionProvider);

        const databaseProvider = new ArangoDatabaseProvider(arangoConnectionProvider);

        testRepositoryProvider = new TestRepositoryProvider(databaseProvider);

        commandInstance = await CommandTestFactory.createTestingCommand({
            imports: [CoscradCliModule],
        })
            .overrideProvider(AppModule)
            .useValue(testAppModule)
            .overrideProvider(REPOSITORY_PROVIDER_TOKEN)
            .useValue(testRepositoryProvider)
            .compile();
    });

    describe(`when the command is invalid`, () => {
        describe(`when the required option "filename" is not provided`, () => {
            /**
             * It's tricky to test when the process exits with status 1. Jest doesn't
             * safely wrap this and the test will fail.
             */
            it.todo(`should fail`);
        });
    });

    describe(`when the command is valid`, () => {
        beforeEach(async () => {
            if (existsSync(outputFilePath)) unlinkSync(outputFilePath);

            await testRepositoryProvider.testTeardown();

            await testRepositoryProvider.addFullSnapshot(
                new DeluxeInMemoryStore(
                    buildTestDataInFlatFormat()
                ).fetchFullSnapshotInLegacyFormat()
            );
        });

        describe(`when using the full --filepath input option`, () => {
            it('should write a dump file', async () => {
                const filePath = `${outputFilePath}_filepath`;

                await CommandTestFactory.run(commandInstance, [
                    'domain-dump',
                    `--filepath=${filePath}`,
                ]);

                const doesFileExist = existsSync(filePath);

                expect(doesFileExist).toBe(true);

                const fileContents = JSON.parse(readFileSync(filePath, { encoding: 'utf-8' }));

                expect(fileContents).toMatchSnapshot();
            });
        });

        describe(`when using the full -f input option`, () => {
            it('should write a dump file', async () => {
                const filePath = `${outputFilePath}_f`;

                await CommandTestFactory.run(commandInstance, ['domain-dump', '-f', filePath]);

                const doesFileExist = existsSync(filePath);

                expect(doesFileExist).toBe(true);

                const fileContents = JSON.parse(readFileSync(filePath, { encoding: 'utf-8' }));

                expect(fileContents).toMatchSnapshot();
            });
        });
    });
});
