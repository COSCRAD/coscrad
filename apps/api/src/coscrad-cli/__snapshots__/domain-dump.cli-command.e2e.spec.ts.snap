// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Command when the command is valid when using the full --filepath input option should write a dump file 1`] = `
[
  {
    "contributorId": "John Doe",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "Chil-term-1",
    "termEnglish": "Engl-term-1",
    "type": "term",
  },
  {
    "contributorId": "John Doe",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "Chil-term-2",
    "termEnglish": "Engl-term-2",
    "type": "term",
  },
  {
    "contributorId": "Jane Deer",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110003",
    "published": false,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "Chil-term-no-english",
    "type": "term",
  },
  {
    "contributorId": "This will be removed soon",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110004",
    "published": false,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [
        "1",
      ],
    },
    "term": "My Secret Term",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110511",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "I am singing (lang)",
    "termEnglish": "I am singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110512",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "You are singing (lang)",
    "termEnglish": "You are singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110513",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "She is singing (lang)",
    "termEnglish": "She is singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110501",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "I am not singing (lang)",
    "termEnglish": "I am not singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110502",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "You are not singing (lang)",
    "termEnglish": "You are not singing (Engl)",
    "type": "term",
  },
]
`;

exports[`Task Command when the command is valid when using the full -f input option should write a dump file 1`] = `
[
  {
    "contributorId": "John Doe",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "Chil-term-1",
    "termEnglish": "Engl-term-1",
    "type": "term",
  },
  {
    "contributorId": "John Doe",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "Chil-term-2",
    "termEnglish": "Engl-term-2",
    "type": "term",
  },
  {
    "contributorId": "Jane Deer",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110003",
    "published": false,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "Chil-term-no-english",
    "type": "term",
  },
  {
    "contributorId": "This will be removed soon",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110004",
    "published": false,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [
        "1",
      ],
    },
    "term": "My Secret Term",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110511",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "I am singing (lang)",
    "termEnglish": "I am singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110512",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "You are singing (lang)",
    "termEnglish": "You are singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110513",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "She is singing (lang)",
    "termEnglish": "She is singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110501",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "I am not singing (lang)",
    "termEnglish": "I am not singing (Engl)",
    "type": "term",
  },
  {
    "contributorId": "Sarah Smith",
    "eventHistory": [],
    "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110502",
    "published": true,
    "queryAccessControlList": {
      "allowedGroupIds": [],
      "allowedUserIds": [],
    },
    "term": "You are not singing (lang)",
    "termEnglish": "You are not singing (Engl)",
    "type": "term",
  },
]
`;
